### PR TITLE
test: expand AutoML and AutoRAG unit test coverage and enforce pipeline resource tiers

### PR DESCRIPTION
**Pull request by LukaszCmielowski**
Assisted-by: Cursor

**Description of your changes:**
# Expand AutoML and AutoRAG unit test coverage and enforce pipeline resource tiers

## Summary

This PR expands unit test coverage for AutoML and AutoRAG components and pipelines, adds CI guardrails for Kubernetes CPU/memory tiers, fixes a timeseries integration test config, adds missing resource constraints to the documents indexing pipeline, and adds cross-artifact parity tests ensuring `pattern.json` and generated notebooks stay in sync.

## Coverage comparison

Line coverage measured with `pytest-cov` on the AutoML/AutoRAG scoped unit suite (`components`, `pipelines`, `utils/tests`; excluding local and integration tests). **Before** = `main`, **After** = this branch.

### Test counts

| Scope | Before (`main`) | After (this PR) | Δ |
|-------|----------------:|----------------:|--:|
| AutoML / AutoRAG scoped tests | 322 | 373 | **+51** |
| Full unit suite | 654 | 705 | **+51** |

### Line coverage by area

| Area | Before | After | Δ |
|------|-------:|------:|--:|
| AutoRAG — documents_discovery | 97.0% | 97.0% | — |
| AutoRAG — documents_indexing | 92.4% | 93.5% | +1.1 |
| AutoRAG — test_data_loader | 90.2% | 92.4% | +2.2 |
| AutoRAG — text_extraction | 79.9% | 81.6% | +1.7 |
| AutoRAG — search_space_preparation | 81.9% | 86.5% | +4.6 |
| AutoRAG — rag_templates_optimization | 83.2% | 83.4% | +0.2 |
| AutoRAG — leaderboard_evaluation | 83.9% | 87.2% | +3.3 |
| AutoRAG — shared (`component_status`) | 86.3% | 86.3% | — |
| AutoML — autogluon_models_training | 91.2% | 91.2% | — |
| AutoML — autogluon_timeseries_models_training | 89.8% | 89.8% | — |
| AutoML — autogluon_leaderboard_evaluation | 96.4% | 96.4% | — |
| Pipeline — documents_indexing | — | 83.3% | **new** |
| Pipeline — rag_optimization | 95.6% | 95.6% | — |
| Pipeline — tabular_training | 94.6% | 94.6% | — |
| Pipeline — timeseries_training | 94.7% | 94.7% | — |
| Utils — `pipeline_task_resources` | — | 87.0% | **new** |
| **Weighted overall (scoped areas)** | **87.2%** | **88.2%** | **+1.0** |

> **—** = not exercised by unit tests on `main` (no pipeline unit tests / file did not exist).

### Component tests

- **AutoRAG search space preparation** — search-space report content, MPS model preselection, notebook-style calls without `component_status`
- **AutoRAG documents indexing** — chunk/embedding parameter forwarding, collection name wiring, batch indexing behavior
- **AutoRAG RAG templates optimization** — pattern artifact JSON (vector store binding), generated indexing/inference notebook substitution
- **AutoRAG leaderboard evaluation** — pattern ranking, optimization-metric column priority, skip empty subdirs, notebook-style calls
- **AutoRAG shared components** — `NullComponentStatusTracker` behavior, JSON schema compliance, notebook-style calls across discovery / text-extraction / test-data-loader
- **AutoML tabular / timeseries training** — balanced preset fit args
- **AutoML leaderboard evaluation** — metric-specific sort order (negated RMSE, MASE)

### Pattern / notebook parity tests *(new)*

`test_pattern_artifacts.py` now includes explicit cross-artifact parity checks (`TestPatternNotebookParity`) that load `pattern.json`, `indexing.ipynb`, and `inference.ipynb` together and assert they match:

- **Vector store binding** — `provider_id` and `vector_store_id` in `pattern.json` match `provider_id` / `collection_name` in both notebooks
- **Indexing notebook ↔ pattern.json** — chunking (`method`, `chunk_size`, `chunk_overlap`) and embedding (`model_id`, `distance_metric`, `embedding_params`)
- **Inference notebook ↔ pattern.json** — retrieval (`method`, `number_of_chunks`, `search_mode`) and generation (`model_id`, prompts)
- **Indexing ↔ inference notebooks** — both reuse the same vector store binding from `pattern.json`
- **S3 keys** — `input_data_key` in indexing notebook, `test_data_key` in inference notebook
- **Multiple patterns** — parity holds independently for each generated pattern directory

### Pipeline tests

- **Documents indexing pipeline** *(new)* — compile, DAG wiring, task dependencies, secret/parameter wiring, resource tier assertions
- **AutoML tabular / timeseries pipelines** — preset-dependent speed vs balanced resource tier assertions
- **RAG optimization pipeline** — component resource tier assertions

### Pipeline resource enforcement

- Add `utils/pipeline_task_resources.py` — compile helpers and expected tier matrices for AutoML and AutoRAG pipelines
- Add `utils/tests/test_pipeline_resource_requirements.py` — assert full executor resource matrices and cross-pipeline alignment for shared steps (`documents-discovery`, `text-extraction`)

### Pipeline fix

- **Documents indexing pipeline** — add missing Kubernetes resource constraints (2 CPU / 8Gi request, 32/64Gi limits) and disable caching on all three tasks, matching the RAG optimization pipeline

### Timeseries test config fix

- Point `timeseries_smoke2` at an existing dataset (`data/timeseries_sales.csv`) and add `known_covariates_names`
- Forward optional `eval_metric` through test config loader; add `test_timeseries_config_loader.py`

## Test plan

- [x] `uv run pytest components pipelines utils/tests --ignore-glob='**/test_*local*.py' --ignore-glob='**/test_*integration*.py' -q` — **705 passed**
- [x] `uv run ruff check` on changed utils, pipeline, and test files — **all checks passed**

**Checklist:**

### Pre-Submission Checklist

- [x] All tests and CI checks pass
- [x] Pre-commit hooks pass without errors
- [x] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [x] The title for your pull request (PR) should follow our title convention.
  [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### Additional Checklist Items for New or Updated Components/Pipelines

- [x] `metadata.yaml` includes fresh `lastVerified` timestamp
- [x] All [required files](https://github.com/kubeflow/pipelines-components/blob/main/docs/CONTRIBUTING.md#required-files)
  are present and complete
- [x] OWNERS file lists appropriate maintainers
- [x] README provides clear documentation with usage examples
- [x] Component follows `snake_case` naming convention
- [x] No security vulnerabilities in dependencies
- [x] Containerfile included if using a custom base image

## Summary by CodeRabbit

* **New Features**
  * Added utilities for compiling and validating Kubernetes executor CPU/memory resource tiers across pipelines.

* **Tests**
  * Expanded test coverage for document indexing, text extraction, and model training components.
  * Added resource requirement validation tests for AutoML and AutoRAG pipelines.
  * Added artifact parity tests for generated notebook templates.

* **Chores**
  * Updated test configurations and pipeline resource specifications for data processing and training workflows.

### CODE REVIEW

File: components/training/automl/autogluon_leaderboard_evaluation/tests/test_leaderboard_metric_sorting.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
+"""Leaderboard sorting tests using real pandas (no DataFrame mocks)."""
```
Comment (Mateusz-Switala) [inline components/training/automl/autogluon_leaderboard_evaluation/tests/test_leaderboard_metric_sorting.py]: Why do we need a separate test file to test sorting models in leaderboard component? It should rather be placed in the existing module `test_component_unit.py`

File: pipelines/training/automl/autogluon_timeseries_training_pipeline/tests/test_timeseries_config_loader.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
+"""Tests for time series integration test config JSON loading."""
```
Comment (Mateusz-Switala) [inline pipelines/training/automl/autogluon_timeseries_training_pipeline/tests/test_timeseries_config_loader.py]: I am not sure if we still need this... The functional tests for automl/autorag are now developed in separate repo: https://github.com/red-hat-data-services/autox-ci/pulls

File: pipelines/training/automl/autogluon_timeseries_training_pipeline/tests/test_configs.json [hunk 3]
```diff
@@ -10,12 +10,13 @@
     "top_n": 2,
     "tags": ["smoke", "timeseries"]
   },
-    {
+  {
     "id": "timeseries_smoke2",
-    "dataset_path": "data/m4_hourly_subset_train.csv",
```
Comment (Mateusz-Switala) [inline pipelines/training/automl/autogluon_timeseries_training_pipeline/tests/test_configs.json:15]: If we no longer uses `data/m4_hourly_subset_train.csv`, we should remove it from the repo

File: utils/tests/test_pipeline_resource_requirements.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
+"""Unit tests for pipeline executor resource helpers and AutoML/AutoRAG tier matrices.
```
Comment (Mateusz-Switala) [inline utils/tests/test_pipeline_resource_requirements.py]: The classes/methods inside this module seems to be AutoML/AutoRAG pipelines specific. Shouldn't be put them in proper test folders where the pipeline are defined? Things in utils should be global, mostly applicable for all pipelines

**Review by Mateusz-Switala — CHANGES_REQUESTED** [review]
## 2. Confirmed Issues & Critical Risks

### Issue 1 — Wrong docstring in `test_speed_preset_fit_args` (factual error in specification)

**File:** `components/training/automl/autogluon_models_training/tests/test_component_unit.py:392`
**Severity: 🔴 High**

The docstring claims:

```python
"""Speed preset uses 1-hour time limit and the original fit args."""
```

The body asserts `45 * 60` (45 minutes), which correctly matches the component source (`PRESET_TIME_LIMITS = {"speed": 45 * 60, ...}`). The docstring is factually wrong. In a test suite, docstrings serve as the specification — a developer reading this would wrongly believe the speed preset should be 60 minutes.

---

### Issue 2 — `inject_component_status` conftest magic hides required parameters

**Files:**
- `components/training/automl/autogluon_models_training/tests/conftest.py`
- `components/training/automl/autogluon_timeseries_models_training/tests/conftest.py`
- `components/training/automl/autogluon_leaderboard_evaluation/tests/conftest.py`

**Severity: 🔴 High (maintainability)**

Three conftest files implement an **identical** autouse wrapper:

```python
@pytest.fixture(autouse=True)
def inject_component_status(monkeypatch, tmp_path):
    from ..component import autogluon_models_training
    original = autogluon_models_training.python_func
    def wrapper(*args, **kwargs):
        if "component_status" not in kwargs:
            art = mock.MagicMock()
            art.path = str(tmp_path / "component_status_out")
            art.metadata = {}
            kwargs["component_status"] = art
        return original(*args, **kwargs)
    monkeypatch.setattr(autogluon_models_training, "python_func", wrapper)
```

This has several problems:

1. **Hidden dependency.** Tests appear to call `python_func` without `component_status`, but silently receive a mocked one. A reader of `test_rejects_empty_label_column` has no idea why the call doesn't fail with a missing-argument error.
2. **Code duplication.** The same pattern is copied verbatim into 3 conftest files.
3. **Zero coverage of the artifact's output.** Because `component_status` is a `MagicMock`, no test verifies that the component actually *writes* anything meaningful to it.
4. **Asymmetry.** `test_leaderboard_metric_sorting.py` explicitly passes `component_status`, while `test_component_unit.py` in the same directory does not. The inconsistency signals the correct pattern was never decided.

---

### Issue 3 — `test_partial_refit_failure_succeeds_with_warnings` asserts an ambiguous contract

**File:** `components/training/automl/autogluon_timeseries_models_training/tests/test_component_unit.py` (~line 590)
**Severity: 🔴 High (logical ambiguity)**

```python
assert result.top_models == ["DeepAR", "TFT", "AutoARIMA"]  # Original top models unchanged
```

TFT's refit failed (`mock_refit_2.fit.side_effect = RuntimeError("OOM")`), yet `top_models` still includes it. A downstream consumer relying on `top_models` to enumerate successfully-built model artifacts would find `TFT_FULL` missing on disk.

The actual source of truth is `metadata["model_names"]`, which the test *does* verify correctly:

```python
assert model_names == ["DeepAR_FULL", "AutoARIMA_FULL"]
```

But the test never explains — or documents — why `top_models` should diverge from `model_names` in the partial-failure case. This is either a bug in the component design or an intentional contract that has no specification.

---

## 3. Maintainability & Readability

### Issue 4 — `from_data_frame.side_effect` chains encode internal call counts

**File:** `components/training/automl/autogluon_timeseries_models_training/tests/test_component_unit.py` (8+ tests)
**Severity: 🟡 Medium**

Every timeseries test manually counts internal `TimeSeriesDataFrame.from_data_frame` calls:

```python
# from_data_frame is called for: train, test, and once per model for build_predict_sample_artifact
mock_ts_df_cls.from_data_frame.side_effect = [_mock_ts_df(), _mock_ts_df(), _mock_ts_df()]
```

Any refactor that adds or removes a single `from_data_frame` call produces a cryptic `StopIteration` from the exhausted `side_effect` list rather than a meaningful assertion error. This pattern repeats across 8+ tests and is a maintenance trap.

---

### Issue 5 — `_executor_resources_from_document` has an ambiguous `or container` fallback

**File:** `utils/pipeline_task_resources.py:67`
**Severity: 🟡 Medium**

```python
raw = container.get("resources") or container
```

If `resources` exists but is an empty dict `{}` (falsy in Python), this silently falls back to the container dict itself. An executor with an empty `resources` block should be **skipped** (no resource declarations), not treated as if its resource keys appear at the container level. This is a logic bug in the edge case.

---

### Issue 9 — Log string assertions are brittle

**File:** `components/training/automl/autogluon_timeseries_models_training/tests/test_component_unit.py` (~line 450)
**Severity: 🟢 Low**

```python
assert "Eval metric 'MASE' was NaN/Inf for model 'DeepAR_FULL'" in caplog.text
assert "Refit failed for model 'DeepAR'" in caplog.text
```

Exact substring matches break on any rewording of log messages. `caplog.records` provides structured access that is more resilient.

---

## 4. Test Coverage Gaps

| Gap | Description |
|---|---|
| **`component_status` output never inspected** | All three conftest fixtures inject a `MagicMock`. No test verifies the component writes meaningful metadata or files to the artifact path. |
| **Timeseries `extra_train_data_path=""` path uncovered** | No test exercises the empty `extra_train_data_path` branch, which skips `from_path`/`concat`. The tabular suite has `test_without_extra_train_data`; timeseries does not. |
| **`test_all_dataset_paths_exist` is a CI time-bomb** | `test_timeseries_config_loader.py:33` asserts physical data files exist on disk. This passes locally if files are committed but fails in any clean environment where test data is absent. |
| **Resource tier constants have no absolute baseline** | The full validation in `utils/tests/test_pipeline_resource_requirements.py` compares constants against themselves. If `TRAINING_SPEED_RESOURCES` is accidentally changed from `("4", …)` to `("2", …)`, all tests still pass — they just redefine what "passing" means. No golden snapshot pins the actual numbers. |

---

## 5. Actionable Fixes

### Fix 1 — Correct the wrong docstring
**Severity: 🔴 High**
**File:** `components/training/automl/autogluon_models_training/tests/test_component_unit.py:392`

```python
# Before
"""Speed preset uses 1-hour time limit and the original fit args."""

# After
"""Speed preset uses a 45-minute time limit and good_quality AutoGluon preset."""
```

---

### Fix 2 — Replace magic conftest injection with explicit fixtures
**Severity: 🔴 High**

Remove the `inject_component_status` autouse wrapper from all three conftest files. Instead, expose a named fixture and pass it explicitly in each test that needs it:

```python
# shared or per-test-module conftest
@pytest.fixture
def component_status_artifact(tmp_path):
    art = mock.MagicMock()
    art.path = str(tmp_path / "component_status_out")
    art.metadata = {}
    return art
```

Then in tests:

```python
def test_regression_happy_path(self, mock_predictor_class, mock_read_csv,
                                component_status_artifact, tmp_path):
    result = autogluon_models_training.python_func(
        ...,
        component_status=component_status_artifact,
    )
```

Additionally, add a dedicated `TestComponentStatusOutput` class to verify what the component actually writes to the artifact.

---

### Fix 3 — Clarify or fix the partial-refit contract
**Severity: 🔴 High**
**File:** `components/training/automl/autogluon_timeseries_models_training/tests/test_component_unit.py` (~line 610)

**If `top_models` intentionally reflects the selection output (not refitted models)**, add a clarifying assertion and comment:

```python
# top_models reflects selection-stage output; metadata["model_names"] is the
# authoritative list of successfully refitted models consumers should use.
assert result.top_models == ["DeepAR", "TFT", "AutoARIMA"]
assert model_names == ["DeepAR_FULL", "AutoARIMA_FULL"]
```

**If this is a bug**, the assertion should be:

```python
assert result.top_models == ["DeepAR", "AutoARIMA"]  # TFT failed refit, excluded
```

---

### Fix 4 — Replace fragile `side_effect` chains with `return_value`
**Severity: 🟡 Medium**
**File:** `components/training/automl/autogluon_timeseries_models_training/tests/test_component_unit.py`

```python
# Before — brittle; breaks if component adds one more from_data_frame call
mock_ts_df_cls.from_data_frame.side_effect = [_mock_ts_df(), _mock_ts_df(), _mock_ts_df()]

# After — resilient; assert specific calls only where behavior matters
mock_ts_df_cls.from_data_frame.return_value = _mock_ts_df()
# Override only where the returned value is load-bearing
mock_ts_df_cls.from_data_frame.side_effect = lambda df, **kw: train_ts if df is train_df else test_ts
```

---

### Fix 5 — Fix the ambiguous `or container` fallback
**Severity: 🟡 Medium**
**File:** `utils/pipeline_task_resources.py:67`

```python
# Before
raw = container.get("resources") or container
if not all(key in raw for key in _RESOURCE_KEYS):
    continue

# After
raw = container.get("resources")
if raw is None or not all(key in raw for key in _RESOURCE_KEYS):
    continue
```

---

### Fix 9 — Use structured `caplog.records` assertions
**Severity: 🟢 Low**
**File:** `components/training/automl/autogluon_timeseries_models_training/tests/test_component_unit.py`

```python
# Before
assert "Eval metric 'MASE' was NaN/Inf for model 'DeepAR_FULL'" in caplog.text

# After — resilient to minor message rewording
error_records = [r for r in caplog.records if r.levelno >= logging.ERROR]
assert any("NaN" in r.message and "DeepAR_FULL" in r.message for r in error_records)
```

---

### Fix 10 — Add `test_without_extra_train_data` for the timeseries component
**Severity: 🟢 Low**

Mirror the tabular `test_without_extra_train_data` test to cover the `extra_train_data_path=""` branch in the timeseries component, verifying `from_path` and `concat` are not called and the component completes successfully.

**Comment by Mateusz-Switala** [comment]
In addition, I believe that for greater clarity and to make reviewer easier, we should split this PR into two: tests for enforce pipeline resource tiers and expand AutoML and AutoRAG tests (much broader topic..., especially for AutoRAG components).

**Comment by LukaszCmielowski** [comment]
Fix 10 — test_without_extra_train_data for timeseries | ⚠️ Not applicable | Timeseries component requires non-empty extra_train_data_path at validation (TypeError if blank). Tabular allows ""; timeseries does not. Adding that test would require a component API change, which is out of scope for this review.
-- | -- | --

**Review by Mateusz-Switala — APPROVED** [review]
/lgtm
